### PR TITLE
Enable both automatic (by the WebsocketClient) and manual (by calling Ping) keep-alives.

### PR DESCRIPTION
**Pull request by enyim**
Ping can also be used to detect if the bot has disconnected.

Not 100% reliable, as some of the exceptions do not "bubble" into Send, but are reported through `WebSocket.OnError`, but supporting them would require a more extensive refactoring of `WebSocketClient`

### CODE REVIEW

**Comment by Workshop2** [comment]
@enyim - I actually need this feature (great work).

I will check it out later it today 👍

File: src/SlackConnector/Connections/Sockets/WebSocketClient.cs [hunk 1]
```diff
@@ -32,18 +33,20 @@ public WebSocketClient(IMessageInterpreter interpreter, string url, ProxySetting
...
-        public async Task Connect()
+
+        public Task Connect()
         {
             var taskSource = new TaskCompletionSource<bool>();
+            EventHandler<ErrorEventArgs> onError = (sender, args) => { taskSource.TrySetException(args.Exception); };
 
-            _webSocket.OnOpen += (sender, args) => { taskSource.SetResult(true); };
-            _webSocket.OnError += (sender, args) => { taskSource.SetException(args.Exception); };
+            _webSocket.OnOpen += (sender, args) => { taskSource.SetResult(true); _webSocket.OnError -= onError; };
```
Comment (Workshop2) [inline src/SlackConnector/Connections/Sockets/WebSocketClient.cs:42]: Can you give context around why you have done this

**Review by Workshop2 — COMMENTED** [review]
Hey @enyim I have a question about the following line of code.

I am just adding some tests around the ping message.

File: src/SlackConnector/Connections/Sockets/WebSocketClient.cs (hunk 1 above)
Comment (enyim) [reply src/SlackConnector/Connections/Sockets/WebSocketClient.cs:42]: Initially, what made me do the Ping impl. was that I could never detect if the bot was disconnected. Forgetting about the catch-all `try...catch`es I found that the code in the constructor **throws an exception**.

Why?

The constructor had a TaskCompletionSource there, that was already Completed (since the connection was successful), but the OnError was still active, and at the first error it tried to set the TSC into a Faulted state -- _even though it was already completed_.

So, I did two things, but any of them would suffice.
1. I unsubscribe from the event after the connection attempt is successful. If not, then there is not need to clean up :)
2. Instead of SetXXX I like to use TrySetXXX on the TSC, as usually it does not matter if the Task is _actually_ transitioned into that state. If a previous attempt already moved it into a different state, well... they won. It's safer for me, as there are no random exceptions, but true, this can hide bugs. 

Here, however, it's safer, as we are working with multi-threaded code of which we have no control over.

What if, 
1. the connection is succesful
2. but then we get disconnected, before we have a chance to unsubscribe from the event? 

With SetXXX: the app will die from some random unhandled (cannot be handled!) exception. 
With TrySet: the app will die (usually) when trying to Say something.

TL;DR:
- TrySet is usually safer than Set on TSCs, as we do not want exceptions in multithreaded code
-  we should always unsubscribe from events we do not use

**Comment by Workshop2** [comment]
Thanks @enyim
